pkg/reconciler/taskrun/taskrun.go

```diff
@@ -667,9 +667,14 @@ func (c *Reconciler) createPod(ctx context.Context, tr *v1beta1.TaskRun, rtr *re
 	// Apply step exitCode path substitution
 	ts = resources.ApplyStepExitCodePath(ts)
 
-	if validateErr := ts.Validate(ctx); validateErr != nil {
-		logger.Errorf("Failed to create a pod for taskrun: %s due to task validation error %v", tr.Name, validateErr)
-		return nil, validateErr
+	if validateWorspacesErr := v1beta1.ValidateDeclaredWorkspaces(ts.Workspaces, ts.Steps, ts.StepTemplate).ViaField("workspaces"); validateWorspacesErr != nil {
```

Comment:
I don't think we should call individual task spec validation sub-functions in the taskRun reconciler. This would require any new validation functions to have two call sites, and it would be easy to forget to add one, allowing bugs to sneak in. In addition, task spec validation covers more than just validation of workspaces. 

An alternative approach could be to modify `taskSpec.Validate` to account for whether any parameters have already been substituted (and only run [validateParameterVariables](https://github.com/tektoncd/pipeline/blob/7059cfd5e8a4f3619353aaf8f2bba722b70afa47/pkg/apis/pipeline/v1beta1/task_validation.go#L275) if they have not yet been substituted).

Reply:
We can use `ctx` for this, similar to what we do for *skipping validation for delete*, etc..

Reply:
I think the idea here was that specifically only the workspace validation be called to validate for a specific issue https://github.com/tektoncd/pipeline/issues/3786 . 
So, I'm not sure every new validation functions need to be called here.  

I do like the idea of taskSpec.Validate knowing if params have been substituted or not by passing it through the context though 👍